packages/@react-aria/combobox/src/useComboBox.ts

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright 2020 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+import {AriaButtonProps} from '@react-types/button';
+import {chain} from '@react-aria/utils';
+import {ComboBoxProps} from '@react-types/combobox';
+import {ComboBoxState} from '@react-stately/combobox';
+import {FocusEvent, HTMLAttributes, RefObject, useEffect, useRef} from 'react';
+import {getItemId, listIds} from '@react-aria/listbox';
+import {ListLayout} from '@react-stately/layout';
+import {useMenuTrigger} from '@react-aria/menu';
+import {useSelectableCollection} from '@react-aria/selection';
+import {useTextField} from '@react-aria/textfield';
+
+interface AriaComboBoxProps<T> extends ComboBoxProps<T> {
+  popoverRef: RefObject<HTMLDivElement>,
+  triggerRef: RefObject<HTMLElement>,
+  inputRef: RefObject<HTMLInputElement>,
+  layout: ListLayout<T>
+}
+
+interface ComboBoxAria {
+  triggerProps: AriaButtonProps,
+  inputProps: HTMLAttributes<HTMLElement>,
+  listBoxProps: HTMLAttributes<HTMLElement>,
+  labelProps: HTMLAttributes<HTMLElement>
+}
+
+export function useComboBox<T>(props: AriaComboBoxProps<T>, state: ComboBoxState<T>): ComboBoxAria {
+  let {
+    triggerRef,
+    popoverRef,
+    inputRef,
+    layout,
+    completionMode = 'suggest',
+    menuTrigger = 'input',
+    allowsCustomValue,
+    onCustomValue
+  } = props;
+
+  let {menuTriggerProps, menuProps} = useMenuTrigger(
+    {
+      type: 'listbox'
+    },
+    state,
+    triggerRef
+  );
+
+  let onChange = (val) => {
+    state.setInputValue(val);
+
+    if (menuTrigger !== 'manual') {
+      state.open();
+    }
+  };
+
+  // Had to set the list id here instead of in useListBox or ListBoxBase so that it would be properly defined
+  // when getting focusedKeyId
+  listIds.set(state, menuProps.id);
```

Comment:
Had to set this listId here rather than in `useListBox` or `ListBoxBase` because we need it for `focusedKeyId` immediately below. I had tried to call `listIds.set` in those two areas above previously but it resulted in `focusedKeyId` being undefined (though why, I couldn't figure out)